File: 2191-sort-the-jumbled-numbers/2191-sort-the-jumbled-numbers.py
class Solution:
    def sortJumbled(self, mapping: List[int], nums: List[int]) -> List[int]:
        # Helper function to get the mapped number
        def getMappedNum(num: int) -> int:
            if num < 10:
                return mapping[num]

            mappedNum = 0
            placeValue = 1

            while num > 0:
                lastDigit = num % 10
                mappedDigit = mapping[lastDigit]
                mappedNum += mappedDigit * placeValue

                placeValue *= 10
                num //= 10

            return mappedNum

        # Create a list of tuples (mapped value, original index)
        mapped_nums = [(getMappedNum(num), i) for i, num in enumerate(nums)]
        
        # Sort based on the mapped values, maintaining original order for ties
        mapped_nums.sort(key=lambda x: (x[0], x[1]))

        # Extract the sorted original numbers based on their mapped values
        return [nums[i] for _, i in mapped_nums]
    

File: 2191-sort-the-jumbled-numbers/2191-sort-the-jumbled-numbers.js
/**
 * @param {number[]} mapping
 * @param {number[]} nums
 * @return {number[]}
 */
var sortJumbled = function(mapping, nums) {
    // Helper function to get the mapped number
    const getMappedNum = (num) => {
        if (num < 10) {
            return mapping[num];
        }

        let mappedNum = 0;
        let placeValue = 1;

        while (num > 0) {
            const lastDigit = num % 10;
            const mappedDigit = mapping[lastDigit];
            mappedNum += mappedDigit * placeValue;

            placeValue *= 10;
            num = Math.floor(num / 10);
        }

        return mappedNum;
    };

    // Create an array of tuples (mapped value, original index)
    const mappedNums = nums.map((num, i) => [getMappedNum(num), i]);

    // Sort based on the mapped values, maintaining original order for ties
    mappedNums.sort((a, b) => {
        if (a[0] === b[0]) {
            return a[1] - b[1];
        }
        return a[0] - b[0];
    });

    // Extract the sorted original numbers based on their mapped values
    return mappedNums.map(([_, i]) => nums[i]);
};
